### PR TITLE
Implement profile retrieval over federation - fixes #651

### LINKED ISSUE #651: Implement profile retrieval over federation
- [ ] Implement [GET /_matrix/federation/v1/query/profile](https://matrix.org/docs/spec/server_server/unstable.html#get-matrix-federation-v1-query-profile)

The scope of this issue is to implement the necessary wiring in Dendrite so that a profile query received on the CS API about a user that's not ours should result in a profile query over federation.

### PR DESCRIPTION
**Pull request by Cnly**
This should fix #651.

Signed-off-by: Alex Chen <minecnly@gmail.com>

### Pull Request Checklist

* [x] I have added any new tests that need to pass to `testfile` as specified in [docs/sytest.md](https://github.com/matrix-org/dendrite/blob/master/docs/sytest.md)
* [x] Pull request includes a [sign off](https://github.com/matrix-org/dendrite/blob/master/CONTRIBUTING.md#sign-off)

### CODE REVIEW

File: clientapi/routing/profile.go [hunk 1]
```diff
@@ -30,49 +30,75 @@ import (
...
-			JSON: jsonerror.NotFound("Bad method"),
-		}
-	}
-	profile, err := appserviceAPI.RetreiveUserProfile(req.Context(), userID, asAPI, accountDB)
+	profile, err := getProfile(req.Context(), accountDB, cfg, userID, asAPI, federation)
 	if err != nil {
+		if err == common.ErrProfileNoExists {
+			return util.JSONResponse{
+				Code: http.StatusNotFound,
+				JSON: jsonerror.NotFound("The user does not exist or does not have a profile"),
```
Comment (anoadragon453) [inline clientapi/routing/profile.go]: Here and the lines below may be a good place to use `common.ErrProfileNoexists`.

File: clientapi/routing/profile.go [hunk 2]
```diff
@@ -30,49 +30,75 @@ import (
...
 func GetAvatarURL(
-	req *http.Request, accountDB *accounts.Database, userID string, asAPI appserviceAPI.AppServiceQueryAPI,
+	req *http.Request, accountDB *accounts.Database, cfg *config.Dendrite,
+	userID string, asAPI appserviceAPI.AppServiceQueryAPI,
+	federation *gomatrixserverlib.FederationClient,
 ) util.JSONResponse {
-	profile, err := appserviceAPI.RetreiveUserProfile(req.Context(), userID, asAPI, accountDB)
+	profile, err := getProfile(req.Context(), accountDB, cfg, userID, asAPI, federation)
 	if err != nil {
+		if err == common.ErrProfileNoExists {
```
Comment (anoadragon453) [inline clientapi/routing/profile.go]: Lines 80-92 are repeated quite a bit and could probably be refactored into their own function.

File: clientapi/routing/profile.go [hunk 3]
```diff
@@ -253,6 +295,38 @@ func SetDisplayName(
 	}
 }
 
+func getProfile(
```
Comment (anoadragon453) [inline clientapi/routing/profile.go]: A docstring (with information about how this function will contact application services/other homeservers) would be handy.

File: clientapi/routing/profile.go (hunk 1 above)
Comment (Cnly) [reply clientapi/routing/profile.go]: Hmmm, not sure I fully get you here. Could you perhaps explain a little bit with the new commit added now?

File: clientapi/routing/profile.go (hunk 2 above)
Comment (Cnly) [reply clientapi/routing/profile.go]: `getProfile` now converts a federation 404 to a `common.ErrProfileNoExists`. Does that seem better now?

Comment (anoadragon453) [reply clientapi/routing/profile.go]: Yes, looks quite clean :)

File: clientapi/routing/profile.go (hunk 1 above)
Comment (anoadragon453) [reply clientapi/routing/profile.go]: Hm, not entirely sure. The current code looks fine though.

**Pull request merged**